src/main/java/apoc/atomic/Atomic.java

```diff
@@ -0,0 +1,250 @@
+package apoc.atomic;
+
+import apoc.atomic.util.AtomicUtils;
+import apoc.util.ArrayBackedList;
+import apoc.util.MapUtil;
+import org.apache.commons.lang3.ArrayUtils;
+import org.neo4j.graphdb.*;
+import org.neo4j.helpers.ArrayUtil;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.procedure.*;
+
+import java.lang.reflect.Array;
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import org.neo4j.helpers.TransactionTemplate;
+
+/**
+ * @author AgileLARUS
+ * @since 20-06-17
+ */
+public class Atomic {
+
+    @Context
+    public GraphDatabaseAPI db;
+
+    /**
+     * increment a property's value
+     */
+    @Procedure(mode = Mode.WRITE)
+    @Description("apoc.atomic.add(node/relatonship,propertyName,number) Sums the property's value with the 'number' value ")
+    public Stream<AtomicResults> add(@Name("container") Object container, @Name("propertyName") String property, @Name("number") Number number, @Name(value = "times", defaultValue = "5") Long times) {
+        checkIsPropertyContainer(container);
+        PropertyContainer propertyContainer;
+        /*AtomicReference to wrap the object*/
+        final AtomicReference newValue = new AtomicReference();
```

Comment:
I think the AtomicReferences are not needed, as those values never leave the method, simple Numbers will do.
The important bit is that it is atomic within neo4j, i.e. taking the write properly lock early on.